### ISSUE TITLE
Failed to generate fail DSN: field value contains \\r\\n

### ISSUE THREAD

**Issue by wjywbs** [foxcpp/maddy#245]
`"smtp_code":550,"smtp_enchcode":"5.2.1","smtp_msg":"gmail-smtp-in.l.google.com. said: The email account that you tried to reach is disabled. Learn more at\n5.2.1  https://support.google.com/mail/?p=DisabledUser - gsmtp"`

`queue: failed to generate fail DSN  {"reason":"failed to write header field #1 (\"Diagnostic-Code\"): field value contains \\r\\n (at 118)"}`

The smtp_msg contains \n.

**Comment by foxcpp** [comment]
The problem is in code generating message/delivery-status (machine-readable) part of DSN message. It abuses go-message header writer due to format similarity and go-message was recently updated to prohibit CR/LF in values (which is a good idea). Error messages with CRLF may come from go-smtp parsing multi-line error responses.

Code that probably needs improving is DSN generator: https://github.com/foxcpp/maddy/blob/master/internal/dsn/dsn.go#L135. I have been pretty busy lately with IRL matters so it is up to you to figure out how to patch it. :)

**Comment by foxcpp** [comment]
CC @emersion 
Are multi-line responses parsed by net/textproto? That 5.2.1 code following \n looks odd so there may also be a problem in go-smtp. Also I wonder if think it might be better to replace \n with " " in go-smtp instead.

**Comment by wjywbs** [comment]
This patch fixes the issue.
```
diff --git a/internal/dsn/dsn.go b/internal/dsn/dsn.go
index b4fc9db..25d3979 100644
--- a/internal/dsn/dsn.go
+++ b/internal/dsn/dsn.go
@@ -7,6 +7,7 @@ import (
        "errors"
        "fmt"
        "io"
+       "strings"
        "text/template"
        "time"
 
@@ -135,7 +136,7 @@ func (info RecipientInfo) WriteTo(utf8 bool, w io.Writer) error {
        if smtpErr, ok := info.DiagnosticCode.(*smtp.SMTPError); ok {
                h.Add("Diagnostic-Code", fmt.Sprintf("smtp; %d %d.%d.%d %s",
                        smtpErr.Code, smtpErr.EnhancedCode[0], smtpErr.EnhancedCode[1], smtpErr.EnhancedCode[2],
-                       smtpErr.Message))
+                       strings.ReplaceAll(strings.ReplaceAll(smtpErr.Message, "\n", " "), "\r", " ")))
        } else if utf8 {
                // It might contain Unicode, so don't include it if we are not allowed to.
                // ... I didn't bother implementing mangling logic to remove Unicode
```

**Comment by foxcpp** [comment]
... Also see emersion/go-smtp#113

**Comment by foxcpp** [comment]
Fix will be included in 0.3.3 release.

**Comment by foxcpp** [comment]
Included in 0.3.3 release.
